### PR TITLE
feat: allow comments in config.json, genesis.json, node_key.json and validator.json

### PR DESCRIPTION
**Pull request by ppca**
This is useful for dynamic config. 
For this to work for node start, nearup also needs to be updated.

### CODE REVIEW

File: core/chain-configs/src/genesis_config.rs [hunk 1]
```diff
@@ -271,7 +272,9 @@ impl GenesisConfig {
     pub fn from_file<P: AsRef<Path>>(path: P) -> anyhow::Result<Self> {
         let file = File::open(path).with_context(|| "Could not open genesis config file.")?;
         let reader = BufReader::new(file);
-        let genesis_config: GenesisConfig = serde_json::from_reader(reader)
+        // Strip the comments from the input (use `as_bytes()` to get a `Read`).
```
Comment (nikurt) [inline core/chain-configs/src/genesis_config.rs:275]:
```suggestion
        // Strip the comments from the input.
        // `as_bytes()` returns a `Read`.
```

File: nearcore/src/dyn_config.rs [hunk 2]
```diff
@@ -59,13 +60,18 @@ where
     for<'a> T: Deserialize<'a>,
 {
     match std::fs::read_to_string(path) {
-        Ok(config_str) => match serde_json::from_str::<T>(&config_str) {
-            Ok(config) => {
-                tracing::info!(target: "neard", config=?config, "Changing the config {path:?}.");
-                return Ok(Some(config));
+        Ok(config_str) => {
+            // Strip the comments from the input (use `as_bytes()` to get a `Read`).
```
Comment (nikurt) [inline nearcore/src/dyn_config.rs:64]: Same as above

**Review by nikurt — COMMENTED** [review]
Should we also allow comments in `node_key.json`, i.e. https://github.com/near/nearcore/blob/master/nearcore/src/config.rs#L1293 ?
Should we also allow comments in `validator_key.json`, i.e. https://github.com/near/nearcore/blob/master/core/primitives/src/validator_signer.rs#L163 ?

> nearup

I don't think that updating nearup is needed. nearup is not the recommended way of running nodes. If comments are not allowed in those conditions, that is acceptable.

**Comment by nikurt** [comment]
> The strip comments for the other json's should also work but not sure how we can test them?

Add comments to `node_key.json` and `validator_key.json`. If the node starts then comments are indeed stripped.

File: core/chain-configs/src/genesis_config.rs [hunk 3]
```diff
@@ -261,18 +261,25 @@ impl GenesisConfig {
     ///
```
Comment (nikurt) [inline core/chain-configs/src/genesis_config.rs:261]: Add a comment that it's not `JSON string`, but `JSON-with-comments string`.

File: nearcore/src/config.rs [hunk 4]
```diff
@@ -395,11 +395,12 @@ impl Default for Config {
 
 impl Config {
     pub fn from_file(path: &Path) -> anyhow::Result<Self> {
```
Comment (nikurt) [inline nearcore/src/config.rs:397]: Comment that the file can be JSON-with-comments.

File: utils/config/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,11 @@
+[package]
```
Comment (nikurt) [inline utils/config/Cargo.toml:1]: Add this in the global `Cargo.toml` as a workspace member. Around line 60.

File: utils/config/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,13 @@
+use std::io::Read;
+
+use json_comments::StripComments;
+
+pub fn strip_comments_from_json_str(json_str: &String) -> std::io::Result<String> {
```
Comment (nikurt) [inline utils/config/src/lib.rs:5]: Good idea to move all the code that deals with `json_comments` to a separate module.
:+1:

**Review by nikurt — APPROVED** [review]

**Comment by iho** [comment]
What motivation for having comments in JSON? Can I remove this feature? With this feature, we have higher startup time. 

https://github.com/near/nearcore/issues/9127#issuecomment-1631000926

**Pull request merged**